### PR TITLE
fix(): return url from listen method

### LINKED ISSUE #875: Return URL from listen method?
:wave: hi there, I'm wondering if you would consider returning either the http instance and/or the address the server is listening on.

Currently doing this:
```js
fastify.listen(process.env.PORT)
  .then(() => console.info(`Server fired up at http://localhost:${process.env.PORT}`));
```

Would be nice to just do:
```js
fastify.listen(process.env.PORT)
  .then(url => console.info(`Server fired up at ${url}`));
```

### PR DESCRIPTION
**Pull request by rjoydip**
Full URL address can be passed in `fastify.listen()`. If listen() is a promise then address passed through resolve method or if it is a callback then passed along with an error. Fixed #875 

Note: Missing "listen on socket" test. Same as previously closed PR #926. Special thanks @allevo, @mcollina, @jsumners, @delvedor 

### Example
```js
fastify.listen(3000, '127.0.0.1')
  .then(address) => /* do whatever */)
  .catch(error) => /* error */);
```
or
```js
fastify.listen(3000, '127.0.0.1', (err, address) => /* do whatever */);
```

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] documentation is changed or added
- [x] tests are included
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by mcollina** [comment]
I think docs should be updated.

Good work!

**Comment by mcollina** [comment]
Can you please remove all calls to close() and add a t.tearDown(fastify.close.bind(fastify)) after creation?

**Comment by rjoydip** [comment]
@mcollina  I have remove all calls to `close()` and added a `t.tearDown(fastify.close.bind(fastify))` after creation and also updated docs.

File: test/listen.test.js [hunk 1]
```diff
@@ -9,47 +9,93 @@ const Fastify = require('..')
...
+
+test('listen accepts a port and a callback with (err, address)', t => {
+  t.plan(3)
+  const fastify = Fastify()
+  t.tearDown(fastify.close.bind(fastify))
+  fastify.listen(0, (err, address) => {
+    fastify.server.unref()
+    t.is(address, 'http://127.0.0.1:' + fastify.server.address().port)
+    t.error(err)
+    t.pass()
```
Comment (mcollina) [inline test/listen.test.js]: there is no need for t.pass here

File: test/listen.test.js [hunk 2]
```diff
@@ -9,47 +9,93 @@ const Fastify = require('..')
...
+})
+
+test('listen accepts a port and a callback with (err, address)', t => {
+  t.plan(3)
+  const fastify = Fastify()
+  t.tearDown(fastify.close.bind(fastify))
+  fastify.listen(0, (err, address) => {
+    fastify.server.unref()
+    t.is(address, 'http://127.0.0.1:' + fastify.server.address().port)
+    t.error(err)
```
Comment (mcollina) [inline test/listen.test.js]: there is no need for t.pass here

File: test/listen.test.js [hunk 3]
```diff
@@ -102,42 +177,111 @@ if (os.platform() !== 'win32') {
...
+  t.plan(1)
+  const fastify = Fastify()
+  t.tearDown(fastify.close.bind(fastify))
+  fastify.listen(undefined)
+    .then((address) => {
+      t.is(address, 'http://127.0.0.1:' + fastify.server.address().port)
+      t.end()
+    })
+    .catch(err => {
+      t.ok(err)
```
Comment (mcollina) [inline test/listen.test.js]: are you sure this is ok?

File: test/listen.test.js [hunk 4]
```diff
@@ -102,42 +177,111 @@ if (os.platform() !== 'win32') {
...
+})
+
+test('listen without port without callback with (address)', t => {
+  t.plan(1)
+  const fastify = Fastify()
+  t.tearDown(fastify.close.bind(fastify))
+  fastify.listen()
+    .then((address) => {
+      t.is(address, 'http://127.0.0.1:' + fastify.server.address().port)
+      t.end()
```
Comment (mcollina) [inline test/listen.test.js]: t.end is not necessart if t.plan is used

Comment (rjoydip) [reply test/listen.test.js]: fixed this one

File: test/listen.test.js (hunk 2 above)
Comment (rjoydip) [reply test/listen.test.js]: also fixed this one

File: test/listen.test.js (hunk 1 above)
Comment (rjoydip) [reply test/listen.test.js]: fixed this one

File: test/listen.test.js (hunk 3 above)
Comment (rjoydip) [reply test/listen.test.js]: Yes sure. I have checked this again.

File: test/listen.test.js [hunk 5]
```diff
@@ -7,49 +7,95 @@ const test = require('tap').test
...
+  t.plan(2)
   const fastify = Fastify()
+  t.tearDown(fastify.close.bind(fastify))
   fastify.listen(0, (err) => {
     fastify.server.unref()
     t.is(fastify.server.address().address, '127.0.0.1')
     t.error(err)
-    t.pass()
-    fastify.close()
+    // t.pass()
```
Comment (mcollina) [inline test/listen.test.js]: can you remove this commented line?

File: test/listen.test.js [hunk 6]
```diff
@@ -102,42 +177,111 @@ if (os.platform() !== 'win32') {
...
 
+test('listen null without callback with (address)', t => {
+  t.plan(1)
+  const fastify = Fastify()
+  t.tearDown(fastify.close.bind(fastify))
+  fastify.listen(null)
+    .then((address) => {
+      t.is(address, 'http://127.0.0.1:' + fastify.server.address().port)
+      t.end()
+    })
```
Comment (mcollina) [inline test/listen.test.js]: These lines are superfluous then

File: test/listen.test.js [hunk 7]
```diff
@@ -102,42 +177,111 @@ if (os.platform() !== 'win32') {
...
+  const fastify = Fastify()
+  t.tearDown(fastify.close.bind(fastify))
+  fastify.listen()
+    .then((address) => {
+      t.is(address, 'http://127.0.0.1:' + fastify.server.address().port)
+      // t.end()
+    })
+    .catch(err => {
+      t.ok(err)
+    })
```
Comment (mcollina) [inline test/listen.test.js]: in this block either .then or .catch are triggered

**Comment by rjoydip** [comment]
Okay. No problem

**Comment by StarpTech** [comment]
@delvedor WDYT about a formatter for our docs?

The easiest way would be to use https://github.com/prettier/prettier for markdown files only.

**Comment by rjoydip** [comment]
Sorry for my previous commit. Can you please check again?

File: fastify.js [hunk 8]
```diff
@@ -281,10 +281,14 @@ function build (options) {
         server.once('error', errEventHandler)
       })
       var listen = new Promise((resolve, reject) => {
-        server.listen(port, address, backlog, () => {
-          server.removeListener('error', errEventHandler)
-          logServerAddress(server.address(), options.https)
-          resolve()
+        server.listen(port, address, backlog, (err) => {
+          if (err) {
```
Comment (delvedor) [inline fastify.js]: There is no need to the listen for the error here, see https://github.com/fastify/fastify/pull/965.

File: test/listen.test.js [hunk 9]
```diff
@@ -7,49 +7,90 @@ const test = require('tap').test
...
   })
 })
 
-test('listen accepts a port, address, and callback', t => {
+test('listen accepts a port and a callback with (err, address)', t => {
   t.plan(2)
   const fastify = Fastify()
+  t.tearDown(fastify.close.bind(fastify))
+  fastify.listen(0, (err, address) => {
+    fastify.server.unref()
```
Comment (delvedor) [inline test/listen.test.js]: There is no need to use both `fastify.server.unref()` and `t.tearDown(fastify.close.bind(fastify))`.

File: test/listen.test.js [hunk 10]
```diff
@@ -7,49 +7,90 @@ const test = require('tap').test
...
+    t.is(address, 'http://127.0.0.1:' + fastify.server.address().port)
+    t.error(err)
+  })
+})
+
+test('listen accepts a port, address, and callback', t => {
+  t.plan(1)
+  const fastify = Fastify()
+  t.tearDown(fastify.close.bind(fastify))
```
Comment (delvedor) [inline test/listen.test.js]: There is no need to use both fastify.server.unref() and t.tearDown(fastify.close.bind(fastify)).

File: test/listen.test.js [hunk 11]
```diff
@@ -7,49 +7,90 @@ const test = require('tap').test
...
+})
+
+test('listen accepts a port, address, backlog and callback with (err, address)', t => {
+  t.plan(2)
+  const fastify = Fastify()
+  t.tearDown(fastify.close.bind(fastify))
+  fastify.listen(0, '127.0.0.1', 511, (err, address) => {
+    fastify.server.unref()
+    t.is(address, 'http://127.0.0.1:' + fastify.server.address().port)
+    t.is(err, null)
```
Comment (delvedor) [inline test/listen.test.js]: Can you use `t.error(err)`?

File: test/listen.test.js [hunk 12]
```diff
@@ -102,42 +179,96 @@ if (os.platform() !== 'win32') {
   test('listen on socket', t => {
     t.plan(2)
     const fastify = Fastify()
+    t.tearDown(fastify.close.bind(fastify))
```
Comment (delvedor) [inline test/listen.test.js]: Why this change?

File: test/listen.test.js [hunk 13]
```diff
@@ -102,42 +179,96 @@ if (os.platform() !== 'win32') {
...
+})
+
+test('listen null without callback with (address)', t => {
+  t.plan(1)
+  const fastify = Fastify()
+  t.tearDown(fastify.close.bind(fastify))
+  fastify.listen(null)
+    .then(address => {
+      t.is(address, 'http://127.0.0.1:' + fastify.server.address().port)
+      t.end()
```
Comment (delvedor) [inline test/listen.test.js]: You don't need `t.end` if you are using `t.plan` :)

File: test/listen.test.js (hunk 10 above)
Comment (rjoydip) [reply test/listen.test.js]: fixed

File: test/listen.test.js (hunk 11 above)
Comment (rjoydip) [reply test/listen.test.js]: fixed

File: test/listen.test.js (hunk 12 above)
Comment (rjoydip) [reply test/listen.test.js]: restore it as previous

File: test/listen.test.js [hunk 14]
```diff
@@ -102,42 +179,96 @@ if (os.platform() !== 'win32') {
...
 
 test('listen without callback', t => {
   t.plan(1)
   const fastify = Fastify()
+  t.tearDown(fastify.close.bind(fastify))
   fastify.listen(0)
     .then(() => {
       t.is(fastify.server.address().address, '127.0.0.1')
-      fastify.close()
+      t.end()
```
[the remark this answers is not in the document]
Comment (rjoydip) [reply test/listen.test.js]: removed `t.end`

File: test/listen.test.js (hunk 13 above)
Comment (rjoydip) [reply test/listen.test.js]: fixed

File: test/listen.test.js [hunk 15]
```diff
@@ -146,39 +277,51 @@ test('double listen without callback rejects', t => {
...
 
 test('listen on invalid port without callback rejects', t => {
   const fastify = Fastify()
-
+  t.tearDown(fastify.close.bind(fastify))
   return fastify.listen(-1)
     .catch(err => {
       t.ok(err)
-      return fastify.close()
+      return true
```
[the remark this answers is not in the document]
Comment (rjoydip) [reply test/listen.test.js]: https://github.com/fastify/fastify/pull/926#discussion_r186267446

File: test/listen.test.js [hunk 16]
```diff
@@ -146,39 +277,51 @@ test('double listen without callback rejects', t => {
...
     })
 })
 
 test('listen logs the port as info', t => {
   t.plan(1)
   const fastify = Fastify()
-
-  t.teardown(() => fastify.close())
+  t.tearDown(fastify.close.bind(fastify))
```
[the remark this answers is not in the document]
Comment (rjoydip) [reply test/listen.test.js]: https://github.com/fastify/fastify/pull/964#issuecomment-395951819

File: test/listen.test.js (hunk 7 above)
Comment (rjoydip) [reply test/listen.test.js]: fixed

File: test/listen.test.js (hunk 6 above)
Comment (rjoydip) [reply test/listen.test.js]: fixed

File: docs/Getting-Started.md [hunk 17]
```diff
@@ -23,11 +24,12 @@ fastify.get('/', function (request, reply) {
 })
 
 // Run the server!
-fastify.listen(3000, function (err) {
+fastify.listen(3000, function (err, address) {
   if (err) {
     fastify.log.error(err)
     process.exit(1)
   }
+  console.log(`server listening on ${address}`)
```
Comment (jsumners) [inline docs/Getting-Started.md]: Use `fastify.log.info` instead of `console.log`.

File: docs/Getting-Started.md [hunk 18]
```diff
@@ -59,11 +61,12 @@ Fastify offers an easy platform that helps solve all of problems, and more.
 > The above examples, and subsequent examples in this document, default to listening *only* on the localhost `127.0.0.1` interface. To listen on all available IPv4 interfaces the example should be modified to listen on `0.0.0.0` like so:
 >
 > ```js
-> fastify.listen(3000, '0.0.0.0', function (err) {
+> fastify.listen(3000, '0.0.0.0', function (err, address) {
 >   if (err) {
 >     fastify.log.error(err)
 >     process.exit(1)
 >   }
+>   console.log(`server listening on ${address}`)
```
Comment (jsumners) [inline docs/Getting-Started.md]: Use `fastify.log.info` instead of `console.log`.

File: docs/Getting-Started.md [hunk 19]
```diff
@@ -81,11 +84,12 @@ const fastify = require('fastify')()
 
 fastify.register(require('./our-first-route'))
 
-fastify.listen(3000, function (err) {
+fastify.listen(3000, function (err, address) {
   if (err) {
     fastify.log.error(err)
     process.exit(1)
   }
+  console.log(`server listening on ${address}`)
```
Comment (jsumners) [inline docs/Getting-Started.md]: Use `fastify.log.info` instead of `console.log`.

File: docs/Getting-Started.md [hunk 20]
```diff
@@ -119,11 +123,12 @@ fastify.register(require('./our-db-connector'), {
 })
 fastify.register(require('./our-first-route'))
 
-fastify.listen(3000, function (err) {
+fastify.listen(3000, function (err, address) {
   if (err) {
     fastify.log.error(err)
     process.exit(1)
   }
+  console.log(`server listening on ${address}`)
```
Comment (jsumners) [inline docs/Getting-Started.md]: Use `fastify.log.info` instead of `console.log`.

File: examples/simple.js [hunk 21]
```diff
@@ -23,7 +23,7 @@ fastify
       .send({ hello: 'world' })
   })
 
-fastify.listen(3000, err => {
+fastify.listen(3000, (err, address) => {
   if (err) throw err
-  console.log(`server listening on ${fastify.server.address().port}`)
+  console.log(`server listening on ${address}`)
```
Comment (jsumners) [inline examples/simple.js]: Use `fastify.log.info` instead of `console.log`.

File: docs/Getting-Started.md (hunk 17 above)
Comment (rjoydip) [reply docs/Getting-Started.md]: fixed

File: docs/Getting-Started.md (hunk 18 above)
Comment (rjoydip) [reply docs/Getting-Started.md]: fixed

File: docs/Getting-Started.md (hunk 19 above)
Comment (rjoydip) [reply docs/Getting-Started.md]: fixed

File: docs/Getting-Started.md (hunk 20 above)
Comment (rjoydip) [reply docs/Getting-Started.md]: fixed

File: examples/simple.js (hunk 21 above)
Comment (rjoydip) [reply examples/simple.js]: fixed
